terps/scott/load_TI99_4a.c

```diff
@@ -0,0 +1,816 @@
+        ap->Condition[0] % 20, ap->Condition[0] / 20);
+    fprintf(stderr, "Action %d Condition[1]: %d (%d/%d)\n", ct, ap->Condition[1],
+        ap->Condition[1] % 20, ap->Condition[1] / 20);
+    fprintf(stderr, "Action %d Condition[2]: %d (%d/%d)\n", ct, ap->Condition[2],
+        ap->Condition[2] % 20, ap->Condition[2] / 20);
+    fprintf(stderr, "Action %d Condition[0]: %d (%d/%d)\n", ct, ap->Condition[3],
+        ap->Condition[3] % 20, ap->Condition[3] / 20);
+    fprintf(stderr, "Action %d Condition[0]: %d (%d/%d)\n", ct, ap->Condition[4],
+        ap->Condition[4] % 20, ap->Condition[4] / 20);
+    fprintf(stderr, "Action %d Action[0]: %d\n", ct, ap->Action[0]);
+    fprintf(stderr, "Action %d Action[1]: %d\n\n", ct, ap->Action[1]);
+}
+
+void print_actions(void)
+{
+    for (int ct = 0; ct <= GameHeader.NumActions; ct++) {
+        print_action(ct);
+    }
+}
+
+uint8_t *LoadTitleScreen(void)
+{
+    char buf[3074];
+    int offset = 0;
+    uint8_t *p;
+    int lines;
+
+    /* title screen offset starts at 0x80 */
+    p = entire_file + 0x80 + file_baseline_offset;
+    if (p - entire_file > file_length)
+        return NULL;
+    for(lines=0; lines<24; lines++)
+    {
+        for (int i = 0; i < 40; i++) {
+            char c = *(p++);
+            if (p - entire_file >= file_length)
+                return NULL;
+            if (!((c <= 127) && (c >= 0))) /* isascii() */
+                c = '?';
+            buf[offset++] = c;
+            if (offset >= 3072)
+                return NULL;
+        }
+        buf[offset++] = '\n';
+    }
+
+    buf[offset] = '\0';
+    uint8_t *result = MemAlloc(offset + 1);
+    memcpy(result, buf, offset + 1);
+    return result;
+}
+
+int try_loading_ti994a(struct DATAHEADER dh, int loud)
+{
+    int ni, na, nw, nr, mc, pr, tr, wl, lt, mn, trm;
+    int ct;
+
+    Room *rp;
+    Item *ip;
+    /* Load the header */
+
+    ni = dh.num_objects;
+    nw = MAX(dh.num_verbs, dh.num_nouns);
+    nr = dh.red_room;
+    mc = dh.max_items_carried;
+    pr = dh.begin_locn;
+    tr = 0;
+    trm = dh.treasure_locn;
+    wl = dh.cmd_length;
+    lt = fix_word(dh.light_turns);
+    mn = max_messages;
+
+    uint8_t *ptr = entire_file;
+
+    GameHeader.NumItems = ni;
+    Items = (Item *)MemAlloc(sizeof(Item) * (ni + 1));
+    na = 400;
+    GameHeader.NumActions = -1;
+    Actions = (Action *)MemAlloc(sizeof(Action) * (na + 1));
+
+    GameHeader.NumWords = nw;
+    GameHeader.WordLength = wl;
+    Verbs = MemAlloc(sizeof(char *) * (nw + 2));
+    Nouns = MemAlloc(sizeof(char *) * (nw + 2));
+    GameHeader.NumRooms = nr;
+    Rooms = (Room *)MemAlloc(sizeof(Room) * (nr + 1));
+    GameHeader.MaxCarry = mc;
+    GameHeader.PlayerRoom = pr;
+    GameHeader.LightTime = lt;
+    LightRefill = lt;
+    GameHeader.NumMessages = mn;
+    Messages = MemAlloc(sizeof(char *) * (mn + 1));
+    GameHeader.TreasureRoom = trm;
+
+    int offset;
+
+#pragma mark rooms
```

Comment:
When warnings are enabled with gcc (which they're not, for some reason, but should be: I must have just forgotten to add it), these pragmas cause warnings.  They ought to be wrapped in tests to see if the compiler is clang.